### PR TITLE
PSMDB-1033 Implement MONGODB-AWS SASL mechanism

### PR DESCRIPTION
**Pull request by igorsol**

### CODE REVIEW

**Review by ktrushin — CHANGES_REQUESTED** [review]

File: src/mongo/db/auth/external/awsiam_server_mechanism.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    try {
+        if (_step == 1) {
+            return _firstStep(inputData);
+        }
+
+        return _secondStep(inputData);
+    } catch (...) {
+        return exceptionToStatus();
+    }
+}
```
[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/auth/external/awsiam_server_mechanism.cpp]: accepted

File: src/mongo/db/auth/external/awsiam_server_mechanism.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    std::vector<char> _serverNonce;
+    std::int32_t _gs2_cb_flag = 0;
+    std::string _userId;
+
+    void appendExtraInfo(BSONObjBuilder* bob) const final;
+    StatusWith<std::tuple<bool, std::string>> stepImpl(OperationContext* opCtx,
+                                                       StringData input) final;
+    StatusWith<std::tuple<bool, std::string>> _firstStep(StringData inputData);
+    StatusWith<std::tuple<bool, std::string>> _secondStep(StringData inputData);
+    void _parseStsResponse(StringData body);
```
[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/auth/external/awsiam_server_mechanism.h]: fixed

[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/auth/external/awsiam_server_mechanism.h]: by the way this header is only included in the corresponding *.cpp file

File: src/mongo/db/auth/external/awsiam_server_mechanism.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    explicit ServerMechanism(std::string authenticationDatabase)
+        : MakeServerMechanism<AWSIAMPolicy>(std::move(authenticationDatabase)) {}
+
+    ~ServerMechanism() final = default;
+
+private:
+    uint32_t _step{0};
+
+    std::vector<char> _serverNonce;
+    std::int32_t _gs2_cb_flag = 0;
```
[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/auth/external/awsiam_server_mechanism.h:70]: fixed

**Review by ktrushin — APPROVED** [review]
Looks good

**Pull request merged**